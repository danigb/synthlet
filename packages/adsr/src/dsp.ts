function createGateDetector() {
  let current = false;
  return (gate: number): boolean | undefined => {
    if (current === false && gate >= 0.9) {
      current = true;
      return true;
    }
    if (current === true && gate < 0.1) {
      current = false;
      return false;
    }
    return undefined;
  };
}

enum Stage {
  Zero,
  Attack,
  Decay,
  Sustain,
  Release,
}

export type AdsrParamInputs = {
  gate: number[];
  attack: number[];
  decay: number[];
  sustain: number[];
  release: number[];
  offset: number[];
  gain: number[];
};

/**
 * An Adsr Generator
 *
 * Based on Nigel Redmon's ADSR code and Will Pirkle's SynthLab Adsr Generator
 *
 * @see https://www.earlevel.com/main/2013/06/01/EG-generators/
 * @see https://github.com/willpirkleaudio/SynthLab/blob/main/source/analogegcore.h
 */
export function createAdsr(sampleRate: number) {
  // TCO values taken from Will Pirkle's SynthLab
  const FADE_IN_TCO = Math.exp(-1.5);
  const FADE_OUT_TCO = Math.exp(-4.95);

  // Params
  let $attack = 0;
  let $decay = 0;
  let $release = 0;
  let $sustain = 0;

  // One pole filters (with b and coefficient)
  const attack = { b: 0, c: 0 };
  const decay = { b: 0, c: 0 };
  const release = { b: 0, c: 0 };

  // Internal state
  const detectGate = createGateDetector();
  let stage: Stage = Stage.Zero;
  let current = 0;

  _updateAdsr(0.01, 0.1, 0.5, 0.3);

  return function adsr(
    input: Float32Array,
    output: Float32Array,
    modifier: boolean,
    params: AdsrParamInputs
  ) {
    _readParams(params);
    const offset = params.offset[0];
    const gain = params.gain[0];

    for (let i = 0; i < output.length; i++) {
      switch (stage) {
        case Stage.Attack:
          current = attack.b + current * attack.c;
          if (current >= 1.0 || $attack <= 0) {
            current = 1.0;
            stage = Stage.Decay;
          }
          break;
        case Stage.Decay:
          current = decay.b + current * decay.c;
          if (current <= $sustain || $decay <= 0) {
            current = $sustain;
            stage = Stage.Sustain;
          }
          break;
        case Stage.Sustain:
          current = $sustain;
          break;
        case Stage.Release:
          current = release.b + current * release.c;
          if (current <= 0.0 || $release <= 0) {
            current = 0.0;
            stage = Stage.Zero;
          }
      }
      const value = modifier ? input[i] * current : current;
      output[i] = value * gain + offset;
    }
  };

  function _readParams(params: AdsrParamInputs) {
    _updateAdsr(
      params.attack[0],
      params.decay[0],
      params.sustain[0],
      params.release[0]
    );

    const gate = detectGate(params.gate[0]);
    if (gate === true) {
      stage = Stage.Attack;
    } else if (gate === false) {
      stage = Stage.Release;
    }
  }

  function _updateAdsr(
    _attack: number,
    _decay: number,
    _sustain: number,
    _release: number
  ) {
    if ($sustain !== _sustain || $decay !== _decay) {
      $sustain = _sustain;
      $decay = _decay;
      _updateFilter(decay, $decay, $sustain, FADE_OUT_TCO);
    }
    if ($attack !== _attack) {
      $attack = _attack;
      _updateFilter(attack, $attack, 1 + 2 * FADE_IN_TCO, FADE_IN_TCO);
    }
    if ($release !== _release) {
      $release = _release;
      _updateFilter(release, $release, 0, FADE_OUT_TCO);
    }
  }

  function _updateFilter(
    filter: { b: number; c: number },
    time: number,
    level: number,
    tco: number
  ) {
    const samples = time * sampleRate;
    filter.c = Math.exp(-Math.log((1.0 + tco) / tco) / samples);
    filter.b = (level - tco) * (1.0 - filter.c);
  }
}
