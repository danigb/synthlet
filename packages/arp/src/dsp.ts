export enum ArpType {
  Random = 0,
}

export enum ArpChord {
  Major = 145,
}

export function createArpeggiator() {
  let $note = 60;
  let $chord = 0;
  let $octaves = 1;

  let chordNotes = [0];
  let position = 0;
  let len = 1;
  let active = false;
  let current = $note;

  return function update(
    trigger: number,
    type: number,
    baseNote: number,
    chord: number,
    octaves: number
  ): number {
    $note = baseNote;
    $octaves = octaves;

    if ($chord !== chord) {
      $chord = chord;
      chordNotes = getPitchClasses(chord);
      len = chordNotes.length;
      position = position % len;
    }

    if (trigger === 1) {
      if (!active) {
        active = true;
        current = getNextNote(type);
      }
    } else {
      active = false;
    }

    return current;
  };

  function getNextNote(type: number) {
    switch (type) {
      case ArpType.Random:
      default:
        return nextRandom();
    }
  }

  function nextRandom() {
    const octave = Math.floor(Math.random() * ($octaves - 1));
    const randomFromChord = chordNotes[Math.floor(Math.random() * len)];
    return $note + randomFromChord + octave * 12;
  }
}

function getPitchClasses(chord: number) {
  const binary = chord.toString(2);
  const pitchClasses: number[] = [];
  for (let i = 0; i < binary.length; i++) {
    if (binary[i] === "1") {
      pitchClasses.push(i);
    }
  }
  return pitchClasses;
}
