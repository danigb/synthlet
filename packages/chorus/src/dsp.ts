export function createChorus(sampleRate: number) {
  const ftbl0ChorusSIG0 = new Float32Array(65536);
  const ftbl1ChorusSIG1 = new Float32Array(65536);

  let $iVec0 = new Int32Array(2);
  let $fVslider0 = 0.0;
  const $fSampleRate = sampleRate;
  let $fConst0 = 0.0;
  let $fConst1 = 0.0;
  let $fConst2 = 0.0;
  let $fVslider1 = 0.0;
  let $fRec0 = new Float32Array(2);
  let $IOTA0 = 0;
  let $fVec1 = new Float32Array(8192);
  let $fVslider2 = 0.0;
  let $fRec1 = new Float32Array(2);
  let $fVslider3 = 0.0;
  let $fRec2 = new Float32Array(2);
  let $fConst3 = 0.0;
  let $fVslider4 = 0.0;
  let $fRec5 = new Float32Array(2);
  let $fRec4 = new Float32Array(2);
  let $fConst4 = 0.0;
  let $fRec7 = new Float32Array(2);
  let $fConst5 = 0.0;
  let $fRec8 = new Float32Array(2);
  let $fConst6 = 0.0;
  let $fRec9 = new Float32Array(2);
  let $fConst7 = 0.0;
  let $fRec10 = new Float32Array(2);
  let $fConst8 = 0.0;
  let $fRec11 = new Float32Array(2);
  let $fConst9 = 0.0;
  let $fRec12 = new Float32Array(2);

  $fConst0 = Math.min(1.92e5, Math.max(1.0, $fSampleRate));
  $fConst1 = Math.exp(-(44.12234 / $fConst0));
  $fConst2 = 1.0 - $fConst1;
  $fConst3 = 0.33333334 / $fConst0;
  $fConst4 = 1.0 / $fConst0;
  $fConst5 = 0.14285715 / $fConst0;
  $fConst6 = 0.5 / $fConst0;
  $fConst7 = 0.25 / $fConst0;
  $fConst8 = 0.16666667 / $fConst0;
  $fConst9 = 0.125 / $fConst0;
  $fVslider0 = 0.5;
  $fVslider1 = 1.0;
  $fVslider2 = 0.5;
  $fVslider3 = 0.5;
  $fVslider4 = 0.5;
  fillTable(ftbl0ChorusSIG0, sig0Fn);
  fillTable(ftbl1ChorusSIG1, sig1Fn);

  function update(
    delay: number,
    rate: number,
    depth: number,
    deviation: number
  ): void {
    $fVslider0 = delay;
    $fVslider1 = 1; // enable
    $fVslider2 = rate;
    $fVslider3 = depth;
    $fVslider4 = deviation;
  }

  function compute(
    inputMono: Float32Array,
    outLeft: Float32Array,
    outRight: Float32Array
  ): void {
    let iSlow0: number = 1 - Math.floor($fVslider0);
    let fSlow1: number = $fConst2 * $fVslider1;
    let fSlow2: number = 4.096 * $fVslider2;
    let fSlow3: number = 6.25e-5 * $fVslider3;
    let fSlow4: number = $fConst2 * $fVslider4;

    for (let i = 0; i < inputMono.length; i++) {
      let input0 = inputMono[i];

      $iVec0[0] = 1;
      $fRec0[0] = fSlow1 + $fConst1 * $fRec0[1];
      let fTemp0: number = input0;
      let fTemp1: number = iSlow0 !== 0 ? 0.0 : fTemp0;
      let fTemp2: number = $fRec0[0] * fTemp1;
      $fVec1[$IOTA0 & 8191] = fTemp2;
      $fRec1[0] = fSlow2 + 0.999 * $fRec1[1];
      $fRec2[0] = fSlow3 * $fRec1[0] + 0.999 * $fRec2[1];

      let iTemp3: number = 1 - $iVec0[1];
      $fRec5[0] = fSlow4 + $fConst1 * $fRec5[1];
      let fTemp4: number =
        iTemp3 !== 0 ? 0.0 : $fRec4[1] + $fConst3 * $fRec5[0];
      $fRec4[0] = fTemp4 - Math.floor(fTemp4);

      let fTemp5: number = Math.min(
        4096.0,
        0.375 * $fRec1[0] +
          $fRec2[0] *
            ftbl0ChorusSIG0[
              Math.max(0, Math.min(Math.floor(65536.0 * $fRec4[0]), 65535))
            ]
      );
      let iTemp6: number = Math.floor(fTemp5);
      let fTemp7: number = Math.floor(fTemp5);

      let fTemp8: number =
        iTemp3 !== 0 ? 0.0 : $fRec7[1] + $fConst4 * $fRec5[0];
      $fRec7[0] = fTemp8 - Math.floor(fTemp8);

      let fTemp9: number = Math.min(
        4096.0,
        0.125 * $fRec1[0] +
          $fRec2[0] *
            ftbl1ChorusSIG1[
              Math.max(0, Math.min(Math.floor(65536.0 * $fRec7[0]), 65535))
            ]
      );
      let fTemp10: number = Math.floor(fTemp9);
      let iTemp11: number = Math.floor(fTemp9);

      let fTemp12: number = fTemp1 * (1.0 - $fRec0[0]);

      let fTemp13: number =
        iTemp3 !== 0 ? 0.0 : $fRec8[1] + $fConst5 * $fRec5[0];
      $fRec8[0] = fTemp13 - Math.floor(fTemp13);

      let fTemp14: number = Math.min(
        4096.0,
        0.875 * $fRec1[0] -
          $fRec2[0] *
            ftbl0ChorusSIG0[
              Math.max(0, Math.min(Math.floor(65536.0 * $fRec8[0]), 65535))
            ]
      );
      let iTemp15: number = Math.floor(fTemp14);
      let fTemp16: number = Math.floor(fTemp14);

      outLeft[i] =
        iSlow0 !== 0
          ? fTemp0
          : 0.70710677 *
              ($fVec1[($IOTA0 - Math.min(4097, Math.max(0, iTemp6))) & 8191] *
                (fTemp7 + (1.0 - fTemp5)) +
                (fTemp5 - fTemp7) *
                  $fVec1[
                    ($IOTA0 - Math.min(4097, Math.max(0, iTemp6 + 1))) & 8191
                  ]) +
            (fTemp9 - fTemp10) *
              $fVec1[
                ($IOTA0 - Math.min(4097, Math.max(0, iTemp11 + 1))) & 8191
              ] +
            fTemp12 +
            $fVec1[($IOTA0 - Math.min(4097, Math.max(0, iTemp11))) & 8191] *
              (fTemp10 + (1.0 - fTemp9)) -
            0.70710677 *
              ($fVec1[($IOTA0 - Math.min(4097, Math.max(0, iTemp15))) & 8191] *
                (fTemp16 + (1.0 - fTemp14)) +
                (fTemp14 - fTemp16) *
                  $fVec1[
                    ($IOTA0 - Math.min(4097, Math.max(0, iTemp15 + 1))) & 8191
                  ]);

      let fTemp17: number =
        iTemp3 !== 0 ? 0.0 : $fRec9[1] + $fConst6 * $fRec5[0];
      $fRec9[0] = fTemp17 - Math.floor(fTemp17);

      let iTemp18: number = Math.max(
        0,
        Math.min(Math.floor(65536.0 * $fRec9[0]), 65535)
      );

      let fTemp19: number = Math.min(
        4096.0,
        0.25 * $fRec1[0] +
          $fRec2[0] *
            (0.70710677 * ftbl1ChorusSIG1[iTemp18] +
              0.70710677 * ftbl0ChorusSIG0[iTemp18])
      );
      let iTemp20: number = Math.floor(fTemp19);
      let fTemp21: number = Math.floor(fTemp19);

      let fTemp22: number =
        iTemp3 !== 0 ? 0.0 : $fRec10[1] + $fConst7 * $fRec5[0];
      $fRec10[0] = fTemp22 - Math.floor(fTemp22);

      let iTemp23: number = Math.max(
        0,
        Math.min(Math.floor(65536.0 * $fRec10[0]), 65535)
      );

      let fTemp24: number = Math.min(
        4096.0,
        0.5 * $fRec1[0] +
          $fRec2[0] *
            (0.70710677 * ftbl0ChorusSIG0[iTemp23] -
              0.70710677 * ftbl1ChorusSIG1[iTemp23])
      );
      let iTemp25: number = Math.floor(fTemp24);
      let fTemp26: number = Math.floor(fTemp24);

      let fTemp27: number =
        iTemp3 !== 0 ? 0.0 : $fRec11[1] + $fConst8 * $fRec5[0];
      $fRec11[0] = fTemp27 - Math.floor(fTemp27);

      let iTemp28: number = Math.max(
        0,
        Math.min(Math.floor(65536.0 * $fRec11[0]), 65535)
      );

      let fTemp29: number = Math.min(
        4096.0,
        0.75 * $fRec1[0] -
          $fRec2[0] *
            (0.70710677 * ftbl1ChorusSIG1[iTemp28] +
              0.70710677 * ftbl0ChorusSIG0[iTemp28])
      );
      let iTemp30: number = Math.floor(fTemp29);
      let fTemp31: number = Math.floor(fTemp29);

      let fTemp32: number =
        iTemp3 !== 0 ? 0.0 : $fRec12[1] + $fConst9 * $fRec5[0];
      $fRec12[0] = fTemp32 - Math.floor(fTemp32);

      let iTemp33: number = Math.max(
        0,
        Math.min(Math.floor(65536.0 * $fRec12[0]), 65535)
      );

      let fTemp34: number = Math.min(
        4096.0,
        $fRec1[0] +
          $fRec2[0] *
            (0.70710677 * ftbl1ChorusSIG1[iTemp33] -
              0.70710677 * ftbl0ChorusSIG0[iTemp33])
      );
      let iTemp35: number = Math.floor(fTemp34);
      let fTemp36: number = Math.floor(fTemp34);

      outRight[i] =
        iSlow0 !== 0
          ? fTemp0
          : fTemp12 -
            (0.38268343 *
              ($fVec1[($IOTA0 - Math.min(4097, Math.max(0, iTemp20))) & 8191] *
                (fTemp21 + (1.0 - fTemp19)) +
                (fTemp19 - fTemp21) *
                  $fVec1[
                    ($IOTA0 - Math.min(4097, Math.max(0, iTemp20 + 1))) & 8191
                  ]) +
              0.9238795 *
                ($fVec1[
                  ($IOTA0 - Math.min(4097, Math.max(0, iTemp25))) & 8191
                ] *
                  (fTemp26 + (1.0 - fTemp24)) +
                  (fTemp24 - fTemp26) *
                    $fVec1[
                      ($IOTA0 - Math.min(4097, Math.max(0, iTemp25 + 1))) & 8191
                    ]) +
              0.9238795 *
                ($fVec1[
                  ($IOTA0 - Math.min(4097, Math.max(0, iTemp30))) & 8191
                ] *
                  (fTemp31 + (1.0 - fTemp29)) +
                  (fTemp29 - fTemp31) *
                    $fVec1[
                      ($IOTA0 - Math.min(4097, Math.max(0, iTemp30 + 1))) & 8191
                    ]) +
              0.38268343 *
                ($fVec1[
                  ($IOTA0 - Math.min(4097, Math.max(0, iTemp35))) & 8191
                ] *
                  (fTemp36 + (1.0 - fTemp34)) +
                  (fTemp34 - fTemp36) *
                    $fVec1[
                      ($IOTA0 - Math.min(4097, Math.max(0, iTemp35 + 1))) & 8191
                    ]));

      $iVec0[1] = $iVec0[0];
      $fRec0[1] = $fRec0[0];
      $IOTA0 = ($IOTA0 + 1) & 8191;
      $fRec1[1] = $fRec1[0];
      $fRec2[1] = $fRec2[0];
      $fRec5[1] = $fRec5[0];
      $fRec4[1] = $fRec4[0];
      $fRec7[1] = $fRec7[0];
      $fRec8[1] = $fRec8[0];
      $fRec9[1] = $fRec9[0];
      $fRec10[1] = $fRec10[0];
      $fRec11[1] = $fRec11[0];
      $fRec12[1] = $fRec12[0];
    }
  }

  return { update, compute };
}

const sig0Fn = (x: number) => Math.cos(9.58738e-5 * x);
const sig1Fn = (x: number) => Math.sin(9.58738e-5 * x);

function fillTable(table: Float32Array, fn: (phase: number) => void) {
  if (table.length !== 65536) {
    throw new Error("Table must be 65536 samples long");
  }
  let iVec2A = 0;
  let iVec2B = 0;
  let iRec3A = 0;
  let iRec3B = 0;
  for (let i1 = 0; i1 < 65536; i1++) {
    iVec2A = 1;
    iRec3A = (iVec2B + iRec3B) % 65536;
    table[i1] = Math.cos(9.58738e-5 * iRec3A);
    iVec2B = iVec2A;
    iRec3B = iRec3A;
  }
  return table;
}
