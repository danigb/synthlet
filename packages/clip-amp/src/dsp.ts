export enum ClipType {
  Tanh = 0,
}

export function getClipFn(type: ClipType): (x: number) => number {
  switch (type) {
    case ClipType.Tanh:
      return tanh;
    default:
      return bypass;
  }
}

const bypass = (x: number) => x;
const tanh = Math.tanh;
