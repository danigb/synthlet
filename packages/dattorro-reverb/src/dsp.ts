export type UpdateFn = ReturnType<typeof createDsp>["update"];
export type ComputeFn = ReturnType<typeof createDsp>["compute"];

// This is a dsp code generated from Faust to test if the Faust -> Rust -> Typescript conversion works
export function createDsp(sampleRate: number) {
  let fConst0 = 44.1 / sampleRate;
  let fConst1 = 1 - fConst0;
  let fVslider0 = -6.0;
  let fVslider1 = 0.0;
  let fVslider2 = 0.7;
  let fVslider3 = 0.625;
  let fVslider4 = 0.625;
  let fVslider5 = 0.7;
  let fVslider6 = 0.625;
  let fVslider7 = 0.625;
  let fVslider8 = 0.625;
  let IOTA0 = 0;
  let fRec0 = [0.0, 0.0];
  let fRec1 = [0.0, 0.0];
  let fRec13 = [0.0, 0.0];
  let fRec12 = [0.0, 0.0, 0.0];
  let fRec14 = [0.0, 0.0];
  let fVec0 = new Array(256).fill(0.0);
  let fRec10 = [0.0, 0.0];
  let fVec1 = new Array(128).fill(0.0);
  let fRec8 = [0.0, 0.0];
  let fRec15 = [0.0, 0.0];
  let fVec2 = new Array(512).fill(0.0);
  let fRec6 = [0.0, 0.0];
  let fVec3 = new Array(512).fill(0.0);
  let fRec4 = [0.0, 0.0];
  let fRec16 = [0.0, 0.0];
  let fRec20 = [0.0, 0.0];
  let fRec23 = [0.0, 0.0];
  let fVec4 = new Array(1024).fill(0.0);
  let fRec21 = [0.0, 0.0];
  let fVec5 = new Array(8192).fill(0.0);
  let fRec19 = [0.0, 0.0];
  let fRec24 = [0.0, 0.0];
  let fVec6 = new Array(4096).fill(0.0);
  let fRec17 = [0.0, 0.0];
  let fVec7 = new Array(4096).fill(0.0);
  let fRec2 = [0.0, 0.0];
  let fVec8 = new Array(1024).fill(0.0);
  let fRec28 = [0.0, 0.0];
  let fVec9 = new Array(8192).fill(0.0);
  let fRec27 = [0.0, 0.0];
  let fVec10 = new Array(2048).fill(0.0);
  let fRec25 = [0.0, 0.0];
  let fVec11 = new Array(2048).fill(0.0);
  let fRec3 = [0.0, 0.0];

  function update(
    lpFilter: number,
    diffussion1: number,
    diffussion2: number,
    feedback1: number,
    feedback2: number,
    decayRate: number,
    damping: number,
    dryWet: number,
    volume: number
  ) {
    fVslider0 = lpFilter;
    fVslider1 = diffussion1;
    fVslider2 = diffussion2;
    fVslider3 = feedback1;
    fVslider4 = feedback2;
    fVslider5 = decayRate;
    fVslider6 = damping;
    fVslider7 = dryWet;
    fVslider8 = volume;
  }

  function compute(
    inputs: Float32Array[],
    outputs: Float32Array[],
    count: number
  ) {
    let inputs0 = inputs[0];
    let inputs1 = inputs.length === 1 ? inputs[0] : inputs[1];
    let outputs0 = outputs[0];
    let outputs1 = outputs[1];

    let fSlow0 = fConst0 * Math.pow(10, 0.05 * fVslider0);
    let fSlow1 = fConst0 * fVslider1;
    let fSlow2 = fConst0 * fVslider2;
    let fSlow3 = fConst0 * fVslider3;
    let fSlow4 = fConst0 * fVslider4;
    let fSlow5 = fConst0 * fVslider5;
    let fSlow6 = fConst0 * fVslider6;
    let fSlow7 = fConst0 * fVslider7;
    let fSlow8 = fConst0 * fVslider8;

    for (let i = 0; i < count; i++) {
      let input0 = inputs0[i];
      let input1 = inputs1[i];

      fRec0[0] = fSlow0 + fConst1 * fRec0[1];
      let fTemp0 = input0;
      fRec1[0] = fSlow1 + fConst1 * fRec1[1];
      let fTemp1 = fRec1[0] + 1.0;
      let fTemp2 = 1.0 - 0.5 * fTemp1;
      fRec13[0] = fSlow2 + fConst1 * fRec13[1];
      let fTemp3 = input1;
      fRec12[0] =
        (1.0 - fRec13[0]) * fRec12[2] + 0.5 * (fTemp0 + fTemp3) * fRec13[0];
      fRec14[0] = fSlow3 + fConst1 * fRec14[1];
      let fTemp4 = fRec12[0] - fRec14[0] * fRec10[1];
      fVec0[IOTA0 & 255] = fTemp4;
      fRec10[0] = fVec0[(IOTA0 - 142) & 255];
      let fRec11 = fRec14[0] * fTemp4;
      let fTemp5 = fRec11 + fRec10[1] - fRec14[0] * fRec8[1];
      fVec1[IOTA0 & 127] = fTemp5;
      fRec8[0] = fVec1[(IOTA0 - 107) & 127];
      let fRec9 = fRec14[0] * fTemp5;
      fRec15[0] = fSlow4 + fConst1 * fRec15[1];
      let fTemp6 = fRec9 + fRec8[1] - fRec15[0] * fRec6[1];
      fVec2[IOTA0 & 511] = fTemp6;
      fRec6[0] = fVec2[(IOTA0 - 379) & 511];
      let fRec7 = fRec15[0] * fTemp6;
      let fTemp7 = fRec7 + fRec6[1] - fRec15[0] * fRec4[1];
      fVec3[IOTA0 & 511] = fTemp7;
      fRec4[0] = fVec3[(IOTA0 - 277) & 511];
      let fRec5 = fRec15[0] * fTemp7;

      fRec16[0] = fSlow5 + fConst1 * fRec16[1];
      fRec20[0] = fSlow6 + fConst1 * fRec20[1];
      let fTemp8 = 1.0 - fRec20[0];
      fRec23[0] = fSlow7 + fConst1 * fRec23[1];
      let fTemp9 = fRec23[0] * fRec21[1] + fRec3[1];
      fVec4[IOTA0 & 1023] = fTemp9;
      fRec21[0] = fVec4[(IOTA0 - 908) & 1023];
      let fRec22 = -(fRec23[0] * fTemp9);
      fVec5[IOTA0 & 8191] = fRec22 + fRec21[1];
      fRec19[0] = fRec20[0] * fRec19[1] + fTemp8 * fVec5[(IOTA0 - 4217) & 8191];
      fRec24[0] = fSlow8 + fConst1 * fRec24[1];
      let fTemp10 = fRec19[0] * fRec16[0] - fRec24[0] * fRec17[1];
      fVec6[IOTA0 & 4095] = fTemp10;
      fRec17[0] = fVec6[(IOTA0 - 2656) & 4095];
      let fRec18 = fRec24[0] * fTemp10;
      fVec7[IOTA0 & 4095] = fRec18 + fRec17[1];
      fRec2[0] = fRec5 + fRec16[0] * fVec7[(IOTA0 - 2656) & 4095] + fRec4[1];
      let fTemp11 = fRec23[0] * fRec28[1] + fRec2[1];
      fVec8[IOTA0 & 1023] = fTemp11;
      fRec28[0] = fVec8[(IOTA0 - 672) & 1023];
      let fRec29 = -(fRec23[0] * fTemp11);
      fVec9[IOTA0 & 8191] = fRec29 + fRec28[1];
      fRec27[0] = fRec20[0] * fRec27[1] + fTemp8 * fVec9[(IOTA0 - 4453) & 8191];
      let fTemp12 = fRec16[0] * fRec27[0] - fRec24[0] * fRec25[1];
      fVec10[IOTA0 & 2047] = fTemp12;
      fRec25[0] = fVec10[(IOTA0 - 1800) & 2047];
      let fRec26 = fRec24[0] * fTemp12;
      fVec11[IOTA0 & 2047] = fRec26 + fRec25[1];
      fRec3[0] = fRec5 + fRec4[1] + fRec16[0] * fVec11[(IOTA0 - 1800) & 2047];

      outputs0[i] = fRec0[0] * (fTemp0 * fTemp2 + 0.5 * fTemp1 * fRec2[0]);
      outputs1[i] = fRec0[0] * (fTemp3 * fTemp2 + 0.5 * fTemp1 * fRec3[0]);

      // Update history
      fRec0[1] = fRec0[0];
      fRec1[1] = fRec1[0];
      fRec13[1] = fRec13[0];
      fRec12[2] = fRec12[1];
      fRec12[1] = fRec12[0];
      fRec14[1] = fRec14[0];
      IOTA0++;
      fRec10[1] = fRec10[0];
      fRec8[1] = fRec8[0];
      fRec15[1] = fRec15[0];
      fRec6[1] = fRec6[0];
      fRec4[1] = fRec4[0];
      fRec16[1] = fRec16[0];
      fRec20[1] = fRec20[0];
      fRec23[1] = fRec23[0];
      fRec21[1] = fRec21[0];
      fRec19[1] = fRec19[0];
      fRec24[1] = fRec24[0];
      fRec17[1] = fRec17[0];
      fRec2[1] = fRec2[0];
      fRec28[1] = fRec28[0];
      fRec27[1] = fRec27[0];
      fRec25[1] = fRec25[0];
      fRec3[1] = fRec3[0];
    }
  }

  return { update, compute };
}
