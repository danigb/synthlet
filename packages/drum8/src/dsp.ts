export function createSine(sampleRate: number) {
  const isampleRate = 1 / sampleRate;
  const pi2 = Math.PI * 2;
  let phase = 0;
  let inc = 0;

  return {
    update(frequency: number) {
      inc = frequency * isampleRate;
    },
    generate(output: Float32Array) {
      for (let i = 0; i < output.length; i++) {
        output[i] = Math.sin(phase * pi2);
        phase += inc;
        if (phase >= 1) {
          phase -= 1;
        }
      }
    },
  };
}

// State Variable Filter (mode = lowpass, highpass, bandpass)
export function createFilter(sampleRate: number, mode: number) {
  let d = 0;
  let a = 0;
  let g1 = 0;
  let z0 = 0;
  let z1 = 0;
  let high = 0;
  let band = 0;
  let low = 0;

  const period = 0.5 / sampleRate;
  const pi2 = 2 * Math.PI;

  return {
    update(frequency: number, resonance: number) {
      const invQ = 1.0 / resonance;
      a = 2.0 * Math.tan(pi2 * period * frequency);
      d = 1.0 / (1.0 + invQ * a + a * a);
      g1 = a + invQ;
    },

    process(input: Float32Array) {
      for (let i = 0; i < input.length; i++) {
        const x = input[i];
        high = (x - g1 * z0 - z1) * d;
        band = a * high + z0;
        low = a * band + z1;
        z0 = a * high + band;
        z1 = a * band + low;
        input[i] = mode === 0 ? low : mode === 1 ? high : mode === 2 ? band : x;
      }
    },
  };
}

// Attack-Hold-Release envelope based on https://paulbatchelor.github.io/sndkit/env/
export function createEnvelope(
  sampleRate: number,
  attack: number,
  release: number,
  hold: number
) {
  const MODE_ZERO = 0;
  const MODE_ATTACK = 1;
  const MODE_HOLD = 2;
  const MODE_RELEASE = 3;
  const EPS = 5e-8;

  // Convert seconds to time constants
  const atk_env = Math.exp(-1.0 / (attack * sampleRate));
  const rel_env = Math.exp(-1.0 / (release * sampleRate));

  // Increment for hold time
  const inc = hold <= 0 ? 1.0 : 1.0 / (hold * sampleRate);

  let timer = 0;
  let mode = MODE_ZERO;
  let prev = 0;

  return {
    update(trigger: number) {
      if (trigger === 1) {
        mode = MODE_ATTACK;
      }
    },
    process(output: Float32Array) {
      for (let i = 0; i < output.length; i++) {
        let out = 0;

        switch (mode) {
          case MODE_ATTACK:
            out = atk_env * prev + (1.0 - atk_env);

            if (out - prev <= EPS) {
              mode = MODE_HOLD;
              timer = 0;
            }

            prev = out;
            break;

          case MODE_HOLD:
            out = prev;
            timer += inc;

            if (timer >= 1.0) {
              mode = MODE_RELEASE;
            }
            break;

          case MODE_RELEASE:
            out = rel_env * prev;
            prev = out;

            if (out <= EPS) {
              mode = MODE_ZERO;
            }
            break;

          case MODE_ZERO:
            out = 0;
            break;

          default:
            break;
        }

        output[i] = out;
      }
    },
  };
}
