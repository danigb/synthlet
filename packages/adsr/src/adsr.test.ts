import { createAdsr } from "./dsp";

describe("Adsr Generator", () => {
  const defaultParams = {
    gate: [0],
    attack: [0.3],
    decay: [0.2],
    sustain: [0.5],
    release: [0.4],
    offset: [0],
    gain: [1],
  };

  describe("ADSR", () => {
    it("process input with adsr adsr", () => {
      const adsr = createAdsr(20);

      const outputGateOn = new Float32Array(20);
      adsr(undefined!, outputGateOn, false, {
        ...defaultParams,
        gate: [1],
      });
      expect(outputGateOn).toMatchSnapshot();

      const outputGateOff = new Float32Array(20);
      adsr(undefined!, outputGateOff, false, {
        ...defaultParams,
        gate: [0],
      });
      expect(outputGateOff).toMatchSnapshot();
    });
  });
});
