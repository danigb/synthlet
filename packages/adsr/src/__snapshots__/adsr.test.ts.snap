// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adsr Generator ADSR process input with adsr adsr 1`] = `
Float32Array [
  0.3020015358924866,
  0.5294362306594849,
  0.7007153630256653,
  0.8297041654586792,
  0.9268444776535034,
  1,
  1,
  0.6397664546966553,
  0.535443902015686,
  0.5052323937416077,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
]
`;

exports[`Adsr Generator ADSR process input with adsr adsr 2`] = `
Float32Array [
  0.2657996118068695,
  0.13976648449897766,
  0.07194272428750992,
  0.03544389456510544,
  0.015802331268787384,
  0.005232374649494886,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
